pkcs11/pkcs11_csr.c

```diff
@@ -0,0 +1,548 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#ifndef WOLFSSL_USER_SETTINGS
+#include <wolfssl/options.h>
+#endif
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/wolfcrypt/wc_pkcs11.h>
+#include <wolfssl/wolfcrypt/asn_public.h>
+#include <wolfssl/wolfcrypt/asn.h>
+#include <wolfssl/wolfcrypt/rsa.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/logging.h>
+
+static WC_RNG rng;
+
+#define LARGE_TEMP_SZ 4096
+#define HEAP_HINT NULL
+
+#if defined(HAVE_ECC) && defined(WOLFSSL_CERT_GEN) && \
+    defined(WOLFSSL_CERT_REQ) && defined(WOLFSSL_CERT_REQ)
+#define ENABLE_PKCS11_CSR
+
+/* Check if the internal asn API's are available */
+#if defined(WOLFSSL_TEST_CERT) || defined(OPENSSL_EXTRA) || \
+    defined(OPENSSL_EXTRA_X509_SMALL)
+    #define HAVE_DECODEDCERT
+#endif
+#endif
+
+
+#ifdef ENABLE_PKCS11_CSR
+
+static const char* keyPemFile = "ecc-key.pem";
+static const char* csrPemFile = "ecc-csr.pem";
+static const char* newCertOutput = "ecc-cert.der";
+static const char* newCertPemFile = "ecc-cert.pem";
+/* TODO: Use real certificate associated with PKCS11 key.
+ * This is used to populate the issuer field */
+static const char* caCertPemFile = "../certgen/ca-ecc-cert.pem";
+
+static int gen_csr(void)
+{
+    int ret;
+    int type = ECC_TYPE;
+    ecc_key ecKey;
+    void* keyPtr = NULL;
+    Cert req;
+    byte der[LARGE_TEMP_SZ];
+    int  derSz;
+#ifdef WOLFSSL_DER_TO_PEM
+    byte pem[LARGE_TEMP_SZ];
+    int  pemSz;
+    FILE* file = NULL;
+#endif
+
+    XMEMSET(der, 0, LARGE_TEMP_SZ);
+#ifdef WOLFSSL_DER_TO_PEM
+    XMEMSET(pem, 0, LARGE_TEMP_SZ);
+#endif
+
+#ifdef HAVE_ECC
+    keyPtr = &ecKey;
+    ret = wc_ecc_init_ex(&ecKey, HEAP_HINT, INVALID_DEVID);
+#endif
+    if (ret != 0) {
+        printf("Key initialization failed: %d\n", ret);
+        goto exit;
+    }
+
+    ret = wc_ecc_make_key_ex(&rng, 32, &ecKey, ECC_SECP256R1);
+    if (ret != 0) {
+        printf("Key generation failed: %d\n", ret);
+        goto exit;
+    }
+
+    ret = wc_EccKeyToDer(&ecKey, der, sizeof(der));
+    if (ret <= 0) {
+        printf("Key To DER failed: %d\n", ret);
+        goto exit;
+    }
+    derSz = ret;
+
+#ifdef WOLFSSL_DER_TO_PEM
+    memset(pem, 0, sizeof(pem));
+    ret = wc_DerToPem(der, derSz, pem, sizeof(pem), ECC_PRIVATEKEY_TYPE);
+    if (ret <= 0) {
+        printf("Key DER to PEM failed: %d\n", ret);
+        goto exit;
+    }
+    pemSz = ret;
+    printf("%s (%d)\n", pem, pemSz);
+
+    printf("Saved Key PEM to \"%s\"\n", keyPemFile);
+    file = fopen(keyPemFile, "wb");
+    if (file) {
+        ret = (int)fwrite(pem, 1, pemSz, file);
+        fclose(file);
+    }
+#endif /* WOLFSSL_DER_TO_PEM */
+
+    ret = wc_InitCert(&req);
+    if (ret != 0) {
+        printf("Init Cert failed: %d\n", ret);
+        goto exit;
+    }
+    strncpy(req.subject.country, "US", CTC_NAME_SIZE);
```

Comment:
Extract the setting of subject to a new function and comment it to say it needs to be replaced.